relayer/chain/ethereum/writer_test.go

```diff
@@ -0,0 +1,206 @@
+	db, err := store.PrepareDatabase(dbConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	beefyMessages := make(chan store.DatabaseCmd, 1)
+	dbLogger := logrus.WithField("database", "Beefy")
+	database := store.NewDatabase(db, beefyMessages, dbLogger)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	eg, ctx := errgroup.WithContext(ctx)
+	// Ensure clean termination upon SIGINT, SIGTERM
+	eg.Go(func() error {
+		notify := make(chan os.Signal, 1)
+		signal.Notify(notify, syscall.SIGINT, syscall.SIGTERM)
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case sig := <-notify:
+			logrus.WithField("signal", sig.String()).Info("Received signal")
+			cancel()
+		}
+		return nil
+	})
+
+	// Start database update loop
+	err = database.Start(ctx, eg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// ------------------------- Set up Ethereum writer -------------------------
+	logger, hook := test.NewNullLogger()
+	log := logger.WithField("chain", "TestChain")
+
+	// Set up config
+	config := ethereum.Config{}
+	config.Endpoint = "ws://localhost:8545/"
+	config.PrivateKey = "4e9444a6efd6d42725a250b650a781da2737ea308c839eaccb0f7f3dbd2fea77"
+	config.Contracts.PolkadotRelayChainBridge = "0x8cF6147918A5CBb672703F879f385036f8793a24"
+	config.Contracts.ValidatorRegistry = "0xB1185EDE04202fE62D38F5db72F71e38Ff3E8305"
+	config.BeefyBlockDelay = 5
+
+	kpEth, err := secp256k1.NewKeypairFromString(config.PrivateKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	econn := ethereum.NewConnection(config.Endpoint, kpEth, log)
+	err = econn.Connect(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ethMessages := make(chan []chain.Message, 1)
+	contracts := make(map[substrateTypes.ChannelID]*inbound.Contract)
+	writer, err := ethereum.NewWriter(&config, econn, database, ethMessages, beefyMessages, contracts, log)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = writer.Start(ctx, eg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// ------------------------- Write Beefy Info to contract -------------------------
+	beefyItem := loadSampleBeefyCommitmentInfo()
+
+	// Send NewSignatureCommitment tx
+	err = writer.WriteNewSignatureCommitment(ctx, &beefyItem, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	entries := hook.AllEntries()
+	txSendEntry := entries[len(entries)-2]
+
+	assert.Equal(t, txSendEntry.Level, logrus.InfoLevel)
+	assert.Equal(t, txSendEntry.Message, "New Signature Commitment transaction submitted")
+
+	txHash := common.HexToHash(txSendEntry.Data["txHash"].(string))
+
+	// Wait for the tx to be confirmed
+	isPending := true
+	for isPending {
+		_, isPending, err = econn.GetClient().TransactionByHash(ctx, txHash)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+
+	// Fetch the tx receipt
+	receipt, err := econn.GetClient().TransactionReceipt(ctx, txHash)
+	if err != nil {
+		log.Fatal(err)
+	}
+	assert.Equal(t, receipt.TxHash, txHash)
+	assert.Equal(t, receipt.Status, uint64(1)) // Successful tx status
```

Comment:
Nice, let's add another test case for completion, but fine to put that into the next PR.